### DISCUSSION TITLE
ProLayout 新版本欢迎试用

### DISCUSSION THREAD

**Discussion by chenshuai2144** [ant-design/pro-components#4945] [General]
查看这里直接访问 https://next-procomponents.ant.design/components/layout

基于 css in js 的方案让我们可以动态的设置布局的主题，以下是支持自定的token 列表。
​

## Token

Token 是一种设计系统的基本元素，可以使用 Token 快速的修改组件库的基础样式。Layout 中可以通过 token 属性来配置这些颜色。

### 默认的 token

默认的 token 部分来自于 antd，可能会有一些改动，以下是 layout 依赖的部分。

| token              | 说明           | 默认值                |
| ------------------ | -------------- | --------------------- |
| primaryColor       | 主色           | `#1677FF`             |
| errorColor         | 错误色         | `#ff4d4f`             |
| successColor       | 成功色         | `#52c41a`             |
| infoColor          | 通知色         | `#1677FF`             |
| warningColor       | 警告色         | `#faad14`             |
| headingColor       | 标题色         | `rgba(0, 0, 0, 0.85)` |
| textColor          | 正文色         | `rgba(0, 0, 0, 0.65)` |
| textColorSecondary | 次级色         | `rgba(0, 0, 0, 0.45)` |
| borderColorBase    | 边框基本色     | `#d9d9d9`             |
| borderColorSplit   | 分隔边框的颜色 | `rgba(0, 0, 0, 0.06)` |
| borderRadiusBase   | 默认圆角       | `4px`                 |

### Layout 的 token

| token                     | 说明                            | 默认值                |
| ------------------------- | ------------------------------- | --------------------- |
| layoutBgColor             | layout 的背景颜色               | `#f0f2f5`             |
| appListIconTextColor      | 跨站点应用的图标颜色            | `#666`                |
| appListIconHoverTextColor | 跨站点应用的图标 hover 颜色     | `rgba(0, 0, 0, 0.65)` |
| appListIconHoverTextColor | 跨站点应用的图标 hover 背景颜色 | `rgba(0, 0, 0, 0.04)` |

### Sider Token

Sider Token 是 侧边菜单的色值，与顶部菜单不同。

| token | 说明 | 默认值 |
| --- | --- | --- |
| menuBackgroundColor | menu 的背景颜色 | `transparent` |
| menuTitleTextColor | sider 的标题字体颜色 | `headingColor` |
| menuItemDividerColor | menuItem 分割线的颜色 | `borderColorSplit` |
| menuSubArrowColor | menuItem 的箭头颜色 | `rgba(0, 0, 0, 0.25)` |
| menuTextColor | menuItem 的字体颜色 | `textColor` |
| menuTextColorSecondary | menu 的二级字体颜色，比如 footer 和 action 的 icon | `textColor` |
| menuSelectedTextColor | menuItem 的选中字体颜色 | `rgb(0,0,0)` |
| menuItemHoverBgColor | menuItem 的 hover 背景颜色 | `rgba(90, 75, 75, 0.03)` |
| menuItemSelectedBgColor | menuItem 的选中背景颜色 | `rgba(0, 0, 0, 0.04)` |
| menuItemCollapsedHoverBgColor | 收起 menuItem 的 hover 背景颜色 | `rgba(0, 145, 255, 0.1)` |
| menuItemCollapsedSelectedBgColor | 收起 menuItem 的选中背景颜色 | `rgba(0, 145, 255, 0.08)` |
| collapsedButtonBgColor | 展开收起按钮背景颜色 | `#fff` |
| collapsedButtonTextColor | 展开收起按钮 hover 字体颜色 | `menuTextColorSecondary` |
| collapsedButtonHoverTextColor | 展开收起按钮 hover 时字体颜色 | `menuTextColor` |

### Header Token

| token | 说明 | 默认值 |
| --- | --- | --- |
| headerTitleColor | sider 的标题字体颜色 | `headingColor` |
| menuTextColor | menuItem 的字体颜色 | `textColor` |
| menuTextColorSecondary | menu 的二级字体颜色，比如 footer 和 action 的 icon | `textColor` |
| menuSelectedTextColor | menuItem 的选中字体颜色 | `rgb(0,0,0)` |
| menuItemHoverBgColor | menuItem 的 hover 背景颜色 | `rgba(90, 75, 75, 0.03)` |
| menuItemSelectedBgColor | menuItem 的选中背景颜色 | `rgba(0, 0, 0, 0.04)` |
| menuItemCollapsedHoverBgColor | 收起 menuItem 的 hover 背景颜色 | `rgba(0, 145, 255, 0.1)` |
| menuItemCollapsedSelectedBgColor | 收起 menuItem 的选中背景颜色 | `rgba(0, 145, 255, 0.08)` |
| rightActionsItemTextColor | 右上角字体颜色 | `textColorSecondary` |
| rightActionsItemHoverBgColor | 右上角选中的 hover 颜色 | `rgba(0, 0, 0, 0.03)` |

### pageContainer Token

| token                      | 说明                             | 默认值        |
| -------------------------- | -------------------------------- | ------------- |
| pageContainerContentMargin | pageContainer 自带的 margin      | `#fff`        |
| pageContainerBgColor       | pageContainer 的背景颜色         | `transparent` |
| pageContainerFixedBgColor  | pageContainer 被固定时的背景颜色 | `#FFF`        |

> 测试版本暂时包含这部分内容，未来还会增加。

## 🎏 如何升级到最新版本？

我们可以使用 next 标签来安装最新版本的 ProLayout。
​

```bash
yarn add @ant-design/pro-layout@next
npm i @ant-design/pro-layout@next --save
```

新版本的 ProLayout 主要是样式上的升级， API 变化不是很大的。
​

![image.png](https://cdn.nlark.com/yuque/0/2022/png/84868/1646109504033-9868511e-3815-45b7-8487-f7fedad47b53.png#clientId=u9730b0e7-2252-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=944&id=u5d46fef2&margin=%5Bobject%20Object%5D&name=image.png&originHeight=1888&originWidth=3104&originalType=binary&ratio=1&rotation=0&showTitle=false&size=900295&status=done&style=none&taskId=u2f696f6b-870c-4b9e-b07f-6db0f5bebae&title=&width=1552)

![image.png](https://cdn.nlark.com/yuque/0/2022/png/84868/1646109590572-84a80263-0ca5-4932-9980-aefd21346fb6.png#clientId=u9730b0e7-2252-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=944&id=u2d6721bf&margin=%5Bobject%20Object%5D&name=image.png&originHeight=1888&originWidth=3104&originalType=binary&ratio=1&rotation=0&showTitle=false&size=924470&status=done&style=none&taskId=uc1a9f6f5-c000-4639-93eb-0d42b22baed&title=&width=1552)

![image.png](https://cdn.nlark.com/yuque/0/2022/png/84868/1646109607763-02c2f5ae-0801-4ebb-b51d-b3b0c4c55e61.png#clientId=u9730b0e7-2252-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=944&id=ud9b07f3b&margin=%5Bobject%20Object%5D&name=image.png&originHeight=1888&originWidth=3104&originalType=binary&ratio=1&rotation=0&showTitle=false&size=645549&status=done&style=none&taskId=u4c3b053d-2e2c-4358-ab95-76cf434b248&title=&width=1552)

- `rightContentRender`修改为 `avatarProps` 和 `actionsRender`,  对头像模式做了默认抽象。
- `layout=sider` 模式下，header 消失，只保留了侧边栏，增大可视面积。
- `layout=sider` 模式下，header 消失后，右侧操作区域修改为左下角。
- 增加了 `appList`，用于进行多站点之前的导航。

	
修改的 API 列表

| 参数 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| avatarProps | layout 的头像设置，不同的 layout 放在不同的位置 | `[AvatarProps](https://ant.design/components/avatar-cn/)` | - |
| actionsRender | 自定义操作列表 | `(layoutProps)=>ReactNode[]` | - |
| appList | 跨站点导航列表 | `{ icon, title, desc:, url }[]` | - |

另外此次的最大改变在于样式的重写，减少了 less 的导入，所以样式覆盖的功能需要重新设置。
​

​

## 🙋‍ 反馈给我们

ProLayout 暂时还在测试模式，并且使用了 css in js 的技术，可能会有一些无法覆盖的角落，如果你在使用中碰到了问题，欢迎来给我们报告。
​

你可以在钉钉群 和 github 中找到我们。
​

![](https://camo.githubusercontent.com/a536ab6a244f900a73baaa79613dbcbddb45d06d3931650b2601c2bb9ebc7ebb/68747470733a2f2f6d646e2e616c697061796f626a656374732e636f6d2f797579616e5f716b306f78682f616674732f696d672f2d782d3252717935776f6b41414141414141414141414141466c393441514272#crop=0&crop=0&crop=1&crop=1&from=url&height=384&id=uygiN&margin=%5Bobject%20Object%5D&originHeight=1087&originWidth=874&originalType=binary&ratio=1&rotation=0&showTitle=false&status=done&style=none&title=&width=309)
​

​

​

**Comment by Dunqing** [comment]
Amazing😆

可以在@umijs/plugin-layout中使用吗？

**Comment by chenshuai2144** [comment]
可以的， @umijs/plugin-layout 会拿用户package.json 的问题

**Comment by leshalv** [comment]
PageContainer  区域颜色搭配上会不会太丑了

**Comment by chenshuai2144** [comment]
> PageContainer 区域颜色搭配上会不会太丑了

做了 token 可以配置，设计师是想让主要内容更吸引注意力

**Comment by aikin-vip** [comment]
有没有一个在线预览的新版本呢

**Comment by Dunqing** [comment]
> 有没有一个在线预览的新版本呢

这里看 https://pro-components-preview-pr-4734.surge.sh/components/layout

相关pr: https://github.com/ant-design/pro-components/pull/4734#issuecomment-1053603616

**Comment by XuJin186** [comment]
> PageContainer 区域颜色搭配上会不会太丑了

的确是不好看

**Comment by lukezzz** [comment]
token headerBgColor 设置无效， header background始终为transparent

**Comment by chenshuai2144** [comment]
> token headerBgColor 设置无效， header background始终为transparent


升级一下最新的版本。

**Comment by lukezzz** [comment]
> 升级一下最新的版本。

升级了下到7.0.1-beta.2，设置 headerBgColor还是无效

**Comment by leshalv** [comment]
pro 大概什么时候换上呀？ @chenshuai2144

**Comment by chenshuai2144** [comment]
@[leshalv](https://github.com/leshalv) 要等antd的新版本发布了

**Comment by wangxin305** [comment]
菜单没有选中效果了
https://6237lu.csb.app/?layout=mix&splitMenus=true

**Comment by chenshuai2144** [comment]
> 菜单没有选中效果了
https://6237lu.csb.app/?layout=mix&splitMenus=true


网站上看起来没问题，你有复现吗

**Comment by Dunqing** [comment]
https://user-images.githubusercontent.com/29533304/157648258-2601fe28-2f3e-489b-a6ba-12b74ed05ee3.mp4

看起来是一个bug

收起后打开闪了一下

**Comment by wangxin305** [comment]
> 

<img width="1109" alt="image" src="https://user-images.githubusercontent.com/39551146/157791506-9fd50fc9-1772-41cc-89c2-21b510b43c43.png">
我打就是这样的，切换菜单，没有选中效果

**Comment by strongyc** [comment]
页面无法预览啊，加载很慢

**Comment by IvenBao** [comment]
能升级一下reactrouter吗？routerv6支持一下呗

**Comment by chenshuai2144** [comment]
> 能升级一下reactrouter吗？routerv6支持一下呗

这个去看看新版本umi

**Comment by IvenBao** [comment]
> > 能升级一下reactrouter吗？routerv6支持一下呗
> 
> 这个去看看新版本umi

好的，谢谢

**Comment by IvenBao** [comment]
![image](https://user-images.githubusercontent.com/17780723/158732672-b2df1556-3f17-4022-a7b8-1c661c402d5e.png)
这个ProLayout中的BaseMenu中用到了这个originalTitle，但是react报了 未知props的警告

**Reply by chenshuai2144** [reply]
这个修了再试试

**Comment by leshalv** [comment]
DefaultFooter 样式啥的还没处理吧。

**Comment by leshalv** [comment]
新版嵌套布局的话，后期是不是可以通过pro-layout插件是不是就可以轻松实现了？

**Comment by kaoding** [comment]
新版本 pro-layout 要等 antd v5 和 umi 4 才能体现效果

**Comment by chenshuai2144** [comment]
@[leshalv](https://github.com/leshalv) 会有解决方案。

**Comment by tankpanv** [comment]
你好，我在移动端点击菜单之后,菜单不会自动收回去。我尝试使用menu的autoClose参数，但是不生效。不知道这个有什么解决方案。
![image](https://user-images.githubusercontent.com/13927744/160358314-5c4958d5-c8f0-4dc5-a8fd-e2993b1f603e.png)

![image](https://user-images.githubusercontent.com/13927744/160358087-aef6ac04-5223-45e2-92c9-a878bb8ca59c.png)

**Reply by chenshuai2144** [reply]
看起来是个bug，老哥来个issue，我跟踪一下

**Reply by chenshuai2144** [reply]
发现是个设计。。故意的

**Comment by leshalv** [comment]
<img width="108" alt="image" src="https://user-images.githubusercontent.com/30397655/162615813-6a343ca4-4786-4205-92d4-a79582bdf6d7.png">
没有菜单，屏幕页面缩小时，出现收缩图标

**Reply by chenshuai2144** [reply]
有详细的吗？

**Reply by leshalv** [reply]
https://ant-design-pro-next-layout.vercel.app/
你用这个老哥的，然后F12屏幕改小，在点击那个按钮，没响应

**Reply by leshalv** [reply]
我这个是，没有登录成功，把菜单渲染弄成了[]，然后缩小，也出现了这个BUG

**Comment by Dunqing** [comment]
https://ant-design-pro-next-layout.vercel.app
搞了个示例看看效果

**Comment by scrapstation** [comment]
移动端有适配计划吗？

**Comment by leshalv** [comment]
https://pro-components-preview-pr-4734.surge.sh/~demos/layout-base/?layout=mix&splitMenus=true

F12缩小屏幕，点击展开按钮，无法打开菜单

**Comment by shaodahong** [comment]
`navTheme="dark"` 好像不生效了

**Comment by leshalv** [comment]
内容区域使用 Tabs，有一个 Tab 为Form并采用固定页脚，另一个Tab 是 Table，切换到Table后，还是展示固定页脚。 @chenshuai2144

**Comment by fgprodigal** [comment]
移动端侧边栏收起后，trigger按钮样式错位且无法点击

**Reply by kaoding** [reply]
https://github.com/ant-design/ant-design-pro/issues/9993

**Comment by skyunBoss** [comment]
splitMenus false时， headerBgColor生效，headerTitleColor不生效

**Comment by drizzlesconsin** [comment]
Token 对应的颜色，放出设计稿 + 标注一下多好啊！更清晰、一目了然。
现在这样对得头疼。

**Comment by drizzlesconsin** [comment]
https://github.com/ant-design/pro-components/issues/4741#issuecomment-1080353089

<img width="495" alt="image" src="https://user-images.githubusercontent.com/69514654/178739713-74e1f2e4-ce37-4003-985a-6ea2369aa671.png">

看这个效果图，submenuitem 文本的 `margin-left: 8px;` 样式，没有打包上。

**Comment by tinystorm** [comment]
token特性正式版本有了吗

**Comment by phoema** [comment]
最新的版本那个收缩按钮能够保持和老版本一样的位置吗，在左下角。
我们页面如果改动了位置，需要各种评审。
而且这个按钮位置没有必要破坏性升级吧。
如果靠自定义写一遍，纯后台程序员有点费劲。

**Comment by drizzlesconsin** [comment]
想要 menu `type: group` 生效，一定要这样么，感觉用起来有点麻烦。
```tsx
const [pathname, setPathname] = useState('/list/xxx');

location={{
    pathname,
}}
menuItemRender={(item, dom) => (
  <div onClick={() => setPathname(item.path || '/welcome')}>
    {dom}
  </div>
)}
```

在[旧版](https://github.com/ant-design/ant-design-pro/blob/v6.0.0-beta.1/src/app.tsx#L48)的 app.tsx layout 里没法用。

**Comment by yangxiaoqiao2016** [comment]


**Reply by netcore-jroger** [reply]
```log
TS2305: Module '"@ant-design/pro-components"' has no exported member 'BasicLayoutProps'.
    1 | import React from 'react';
  > 2 | import { BasicLayoutProps, MenuDataItem } from '@ant-design/pro-components';
      |          ^^^^^^^^^^^^^^^^
```

@ant-design/pro-components:2.0.4

**Comment by mushroomlb** [comment]
一级菜单个数超过页面能显示的数量后没有增加省略号显示，而是覆盖了头像区域

**Comment by daifuyang** [comment]
token Sider Token 缺少 menuItem 的选中字体颜色

**Comment by arifszn** [comment]
In my case, the token is not behaving dynamically. Is there any way to achieve dynamic dark mode for the Pro Layout? 

Please see my https://github.com/ant-design/pro-components/issues/6056#issuecomment-1278520925.
